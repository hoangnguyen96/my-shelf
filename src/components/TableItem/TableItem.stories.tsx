import { Meta, StoryFn } from "@storybook/react";
import { ChakraProvider } from "@chakra-ui/react";
import theme from "@app/themes";
import TableItem from ".";

const meta: Meta<typeof TableItem> = {
  component: TableItem,
  decorators: [
    (Story: StoryFn) => (
      <ChakraProvider theme={theme}>
        <Story />
      </ChakraProvider>
    ),
  ],
};

export default meta;

const Template: StoryFn<typeof TableItem> = (args) => <TableItem {...args} />;

export const Default = Template.bind({});
Default.args = { isContribute: true };
