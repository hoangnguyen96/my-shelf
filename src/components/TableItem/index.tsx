"use client";

import { memo } from "react";
import Image from "next/image";
import { Flex, Text, useDisclosure } from "@chakra-ui/react";
import { useRouter } from "next/navigation";
import { ROUTES } from "@app/constants";
import { deleteBook } from "@app/features/dashboard/actions";
import ModalDeleteBook from "../ModalDeleteBook";
import { Button } from "../common";
import HeartIcon from "../HeartIcon";
import StatusBook from "../StatusBook";

interface TableItemProps {
  id: string;
  title: string;
  author: string;
  imageUrl: string;
  category: string;
  edition: string;
  publicationYear: number;
  rating: number;
  isFavorite?: boolean;
  status?: boolean;
  isContribute?: boolean;
  onUpdateFavorites?: () => void;
}

const TableItem = ({
  id,
  title,
  author,
  imageUrl,
  edition,
  category,
  isFavorite = false,
  publicationYear,
  rating,
  status = false,
  isContribute = false,
  onUpdateFavorites,
}: TableItemProps) => {
  const { isOpen, onOpen, onClose } = useDisclosure();
  const router = useRouter();

  const handleRedirectPreview = () => {
    router.push(`${isContribute ? ROUTES.CONTRIBUTE : ROUTES.PREVIEW}/${id}`);
  };

  const handleDeleteBook = () => {
    return onOpen();
  };

  const confirmDeleteBook = async () => {
    await deleteBook(id);
    onClose();
    return router.refresh();
  };

  return (
    <>
      <Flex
        bgColor="white"
        borderRadius="10px"
        boxShadow="0 0 5px 0px rgb(0 0 0 / 10%)"
        py="12px"
        px="24px"
        gap="90px"
        alignItems="center"
      >
        <Flex gap="47px" alignItems="center" maxW={329} w="100%">
          <Image
            src={imageUrl || ""}
            alt="Don't make me think"
            width={75}
            height={100}
            priority
          />
          <Flex flexDir="column">
            <Text size="xl" mb="20px">
              {title}
            </Text>
            <Flex mb="10px">
              <Text size="md">{author},</Text>
              <Text size="md">{publicationYear}</Text>
            </Flex>
            <Text size="xs">{edition} Edition</Text>
          </Flex>
        </Flex>
        <Flex gap="72px" alignItems="center" maxW={312} w="100%">
          <Text size="xl" w="60px">
            {rating}
            <Text as="span" size="md" color="dark.70">
              /5
            </Text>
          </Text>
          <Flex flexDir="column" gap="11px">
            <Text size="xl">{category}</Text>
            <Text size="md">UX Design</Text>
          </Flex>
        </Flex>
        {!isContribute && <StatusBook status={status} />}
        <Flex
          gap={isContribute ? "36px" : "64px"}
          alignItems="center"
          maxW={179}
          w="100%"
        >
          {!isContribute && (
            <HeartIcon
              id={id}
              data-testid="handle-favorite"
              isFavorite={isFavorite}
              onUpdateFavorites={onUpdateFavorites}
            />
          )}
          <Button
            data-testid="redirect-preview"
            size="sm"
            variant="outline"
            text="Preview"
            onClick={handleRedirectPreview}
          />
          {isContribute && (
            <Button
              data-testid="delete-book"
              size="sm"
              variant="outline"
              text="Delete"
              onClick={handleDeleteBook}
            />
          )}
        </Flex>
      </Flex>
      <ModalDeleteBook
        isOpen={isOpen}
        onClose={onClose}
        onConfirm={confirmDeleteBook}
      />
    </>
  );
};

export default memo(TableItem);
