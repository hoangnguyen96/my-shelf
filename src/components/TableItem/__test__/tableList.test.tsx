import "@testing-library/jest-dom";
import { fireEvent, render } from "@testing-library/react";
import TableItem from "..";
import { useRouter } from "next/navigation";

// Mock useRouter
jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
}));

// Mock Image component
jest.mock("next/image", () => ({
  __esModule: true,
  default: (props: any) => <img {...props} />,
}));

describe("TableItem", () => {
  const mockPush = jest.fn();

  beforeEach(() => {
    (useRouter as jest.Mock).mockReturnValue({
      push: mockPush,
    });
  });

  it("Should render correctly snapshot", () => {
    expect(
      render(
        <TableItem
          id={"1"}
          title={"Test"}
          author={"Test"}
          imageUrl={"Test"}
          category={"Test"}
          edition={"Test"}
          publicationYear={0}
          rating={0}
          isFavorite={false}
        />
      )
    ).toMatchSnapshot();
  });

  it("Should render correctly snapshot when imageUrl is empty", () => {
    expect(
      render(
        <TableItem
          id=""
          title=""
          author=""
          imageUrl=""
          category=""
          edition=""
          publicationYear={0}
          rating={0}
          isFavorite={false}
        />
      )
    ).toMatchSnapshot();
  });

  it("Should render correctly snapshot when is contribute", () => {
    expect(
      render(
        <TableItem
          id="1"
          title="Test"
          author="Test"
          imageUrl=""
          category="Test"
          edition="Test"
          isContribute={true}
          publicationYear={0}
          rating={0}
          isFavorite={false}
        />
      )
    ).toMatchSnapshot();
  });

  it("Should handle redirect Preview", () => {
    const { getByTestId } = render(
      <TableItem
        id="1"
        title="Test"
        author="Test"
        imageUrl=""
        category="Test"
        edition="Test"
        isContribute={true}
        publicationYear={0}
        rating={0}
        isFavorite={false}
      />
    );

    const button = getByTestId("redirect-preview");

    fireEvent.click(button);
  });

  it("Should handle redirect Preview when contribute false", () => {
    const { getByTestId } = render(
      <TableItem
        id="1"
        title="Test"
        author="Test"
        imageUrl=""
        category="Test"
        edition="Test"
        isContribute={false}
        publicationYear={0}
        rating={0}
        isFavorite={false}
      />
    );

    const button = getByTestId("redirect-preview");

    fireEvent.click(button);
  });

  it("Should handle delete book", () => {
    const { getByTestId } = render(
      <TableItem
        id="1"
        title="Test"
        author="Test"
        imageUrl=""
        category="Test"
        edition="Test"
        isContribute={true}
        publicationYear={0}
        rating={0}
        isFavorite={false}
      />
    );

    const button = getByTestId("delete-book");

    fireEvent.click(button);
  });
});
