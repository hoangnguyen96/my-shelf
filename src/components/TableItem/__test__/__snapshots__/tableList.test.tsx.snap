// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableItem Should render correctly snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="css-i20yi3"
      >
        <div
          class="css-1jxcnq9"
        >
          <img
            alt="Don't make me think"
            height="100"
            src="Test"
            width="75"
          />
          <div
            class="css-j7qwjs"
          >
            <p
              class="chakra-text css-uewl2b"
            >
              Test
            </p>
            <div
              class="css-1709kwd"
            >
              <p
                class="chakra-text css-0"
              >
                Test
                ,
              </p>
              <p
                class="chakra-text css-0"
              >
                0
              </p>
            </div>
            <p
              class="chakra-text css-0"
            >
              Test
               Edition
            </p>
          </div>
        </div>
        <div
          class="css-gsip0b"
        >
          <p
            class="chakra-text css-q2ms4q"
          >
            0
            <span
              class="chakra-text css-c5z1pm"
            >
              /5
            </span>
          </p>
          <div
            class="css-iqbc30"
          >
            <p
              class="chakra-text css-0"
            >
              Test
            </p>
            <p
              class="chakra-text css-0"
            >
              UX Design
            </p>
          </div>
        </div>
        <div
          class="css-8ji7z5"
        >
          <p
            class="chakra-text css-1619v3u"
          >
            none
          </p>
        </div>
        <div
          class="css-1l4iwxf"
        >
          <button
            aria-label="Heart icon"
            class="chakra-button css-38t1hi"
            data-testid="handle-favorite"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="chakra-icon css-1k8v1e7"
              focusable="false"
              stroke-width="2"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
              />
            </svg>
          </button>
          <button
            class="chakra-button css-4xx2wk"
            type="button"
          >
            Preview
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="css-i20yi3"
    >
      <div
        class="css-1jxcnq9"
      >
        <img
          alt="Don't make me think"
          height="100"
          src="Test"
          width="75"
        />
        <div
          class="css-j7qwjs"
        >
          <p
            class="chakra-text css-uewl2b"
          >
            Test
          </p>
          <div
            class="css-1709kwd"
          >
            <p
              class="chakra-text css-0"
            >
              Test
              ,
            </p>
            <p
              class="chakra-text css-0"
            >
              0
            </p>
          </div>
          <p
            class="chakra-text css-0"
          >
            Test
             Edition
          </p>
        </div>
      </div>
      <div
        class="css-gsip0b"
      >
        <p
          class="chakra-text css-q2ms4q"
        >
          0
          <span
            class="chakra-text css-c5z1pm"
          >
            /5
          </span>
        </p>
        <div
          class="css-iqbc30"
        >
          <p
            class="chakra-text css-0"
          >
            Test
          </p>
          <p
            class="chakra-text css-0"
          >
            UX Design
          </p>
        </div>
      </div>
      <div
        class="css-8ji7z5"
      >
        <p
          class="chakra-text css-1619v3u"
        >
          none
        </p>
      </div>
      <div
        class="css-1l4iwxf"
      >
        <button
          aria-label="Heart icon"
          class="chakra-button css-38t1hi"
          data-testid="handle-favorite"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="chakra-icon css-1k8v1e7"
            focusable="false"
            stroke-width="2"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
            />
          </svg>
        </button>
        <button
          class="chakra-button css-4xx2wk"
          type="button"
        >
          Preview
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
